js/shared.js

```diff
@@ -1438,38 +1438,63 @@ export const swapKeysAndValues = (obj) => {
 };
 
 export const showAnimation = () => {
-    if(document.getElementById('loadingAnimation')) document.getElementById('loadingAnimation').style.display = '';
+    if (document.getElementById('loadingAnimation')) document.getElementById('loadingAnimation').style.display = '';
 }
 
 export const hideAnimation = () => {
-    if(document.getElementById('loadingAnimation')) document.getElementById('loadingAnimation').style.display = 'none';
+    if (document.getElementById('loadingAnimation')) document.getElementById('loadingAnimation').style.display = 'none';
 }
 
 export const subscribeForNotifications = async (data) => {
     const idToken = await getIdToken();
     const response = await fetch(`${api}?api=subscribeToNotification`, {
         method: "POST",
         headers: {
-            Authorization:"Bearer "+idToken,
+            Authorization: "Bearer " + idToken,
             "Content-Type": "application/json"
         },
         body: JSON.stringify(data)
     });
     return await response;
 }
 
-export const retrieveNotifications = async () => {
-  const idToken = await getIdToken();
-  const response = await fetch(`${api}?api=retrieveNotifications`, {
-    method: "GET",
-    headers: {
-      Authorization: "Bearer " + idToken
-    },
-  });
+let processingRetrieveNotifications = false;
+const retrieveNotificationsEventTarget = new EventTarget();
+export const retrieveNotifications = async (markAsRead) => {
+    if (processingRetrieveNotifications) {
+        await waitForRetrieveNotifications();
+    }
+    
+    processingRetrieveNotifications = true;
+    let resultJson = { data: [] };
+    try {
+        const idToken = await getIdToken();
+        const response = await fetch(`${api}?api=retrieveNotifications&markasread=${markAsRead ? 'true' : 'false'}`, {
+            method: "GET",
+            headers: {
+                Authorization: "Bearer " + idToken
+            },
+        });
+        
+        resultJson = await response.json();
 
-  return await response.json();
+    } catch (error) {
+        console.error(error);
+    } finally {
+        processingRetrieveNotifications = false;
+        retrieveNotificationsEventTarget.dispatchEvent(new CustomEvent('done'));
+    }
+    return resultJson;
 };
 
+const waitForRetrieveNotifications = () => {
+    return new Promise((resolve) => {
+        retrieveNotificationsEventTarget.addEventListener('done', () => {
+            resolve();
+        }, {once: true});
+    })
+}
```

Comment:
I like your implementation to prevent the race condition using the `processingRetrieveNotifications` variable as a flag along with the custom event listener here that will resolve in the `finally` block.